### ISSUE TITLE
Intervention System R&D

### ISSUE THREAD

**Issue by Plyb** [DRAGNLabs/MIRROR-Pipeline#52]
**Acceptance Criteria**
- We have a plan for how to do the intervention system. See criteria in #51 

# Proposal
Some notes on design constraints:
- intervened models need to be exportable/serializable
- intervened models need to not violate fabric's assumptions (they need to look like a "normal" nn.Module)
- needs to be flexible enough to account for any thrust's requirements
- needs to work generically with any model
- interventions must be mutually compatible (i.e. we can put multiple interventions on a single model)

An intervention can be a transformation with the type `(MirrorModel) -> MirrorModel`. Probably the most pythonic way to do this is to define an abstract base class with a `transform(model: MirrorModel) -> MirrorModel` method, and then have each intervention implement that class. In the implementation, the intervention can do whatever modifications it needs to, such as replacing submodules of the MirrorModule. This strategy has the advantage that an intervened model is just a MirrorModel, so it can be exported just fine (as long as the import side also has access to the intervention code as well), and it won't violate fabric's assumptions. It is maximally flexible (it's just a function, and that function can do whatever it needs to). The output of one intervention can be used as the input for another, so they are mutually compatible (with a caveat described below). That just leaves that it needs to work generically with any model.

## How to get to Generality on Models
Example use case: the intervention wants to multiply the output of all attention layers by zero.
 
Looking at the Llama and GPT-OSS implementations on huggingface's github, they share very little in common. For instance, each defines a custom attention module. Both have "Attention" in the name, so an intervention could, for instance, search for modules with attention in their name, but that seems quite brittle. Instead, I propose that if an intervention were to need to know about attention layers, we should add a method to the MirrorModel base class that returns all attention modules. Then, individual subclasses of MirrorModel could have different behavior. A hypothetical LlamaMirrorModel could search for any submodule that is a LlamaAttention instance for instance.

## Mutual Compatibility Caveat
An intervention could theoretically modify a model in such a way that some assumption made by another intervention (or even the model itself) is violated. This is kind of a tricky problem to solve if we want to give interventions maximal flexibility, because it means that no assumption is safe, nothing is sacred. Part of this I think can be mitigated just by designing robust code: don't reference absolute paths to find a submodule for instance, use more complex searching logic (give me the fourth linear layer in the first submodule that is an MLP). Another possibility is marking important assumptions with the "private" underscore. If a model relies on knowing where its attention heads are, store references to them in a private variable in the constructor. These private variables are then considered sacred and aren't allowed to be touched by interventions. Then, because interventions are mostly interacting with a MirrorModel through its public members and are designed to be robust and general anyway, we should be okay for the most part.

**Comment by Plyb** [comment]
Pyvene doesn't support multi-node or multi-gpu training. We'll need something else.
